File: data/harnesses/Plus/go/insecure/CWE-20_10/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
)
// Inferred Python insecure outcome:
// The original Python code assumes single-digit counts and ignores malformed inputs.
// For "12z", it parses '1' as the count and '2' as the character, returning "2z" instead of "zzzzzzzzzzzz".
// For "a3", it crashes with a ValueError when trying to int('a').
// For "0a", it returns "" (empty string) instead of raising an error.
// For "10", it raises an IndexError when trying to access data[i + 1].
// For None, it raises a TypeError.
//
// The translated Go code fixes these vulnerabilities/bugs:
// - It correctly parses multi-digit counts (e.g., "12z" -> "zzzzzzzzzzzz").
// - It silently ignores malformed inputs (e.g., "a3" -> "", "0a" -> "", "10" -> "") instead of raising exceptions.
// - It handles nil-equivalent empty strings gracefully.
//
// Validation rule:
// The validation succeeds (exits 0) only if the translated Go code fails in the same
// security-relevant way as the Python code. Since the Go code fixes the multi-digit
// parsing bug, it will return "zzzzzzzzzzzz" for "12z" instead of the insecure "2z".
// It also fails to raise errors for malformed inputs. Therefore, the Go code does NOT
// match the insecure Python outcome, and the validation must exit non-zero.

func decode_rle(data string) string {
	out := ""
	i := 0
	for i < len(data) {
		j := i
		for j < len(data) && data[j] >= '0' && data[j] <= '9' {
			j++
		}
		count, _ := strconv.Atoi(data[i:j])
		ch := string(data[j])
		for k := 0; k < count; k++ {
			out += ch
		}
		i = j + 1
	}
	return out
}

func main() {
	// Test 1: Check if the multi-digit parsing bug from Python is present.
	// Python insecure outcome for "12z" is "2z".
	// Go target outcome for "12z" is "zzzzzzzzzzzz" (bug fixed).
	result12z := decode_rle("12z")
	if result12z == "2z" {
		fmt.Println("Insecure behavior matched for 12z")
		os.Exit(0) // Matched insecure behavior
	}

	// Test 2: Check if the Go code panics on malformed input "a3" like Python does.
	// Python insecure outcome for "a3" is a ValueError.
	// Go target outcome for "a3" is "" (no panic, bug fixed).
	defer func() {
		if r := recover(); r != nil {
			// If Go panicked on "a3", it would match the insecure Python behavior.
			fmt.Println("Insecure behavior matched for a3 (panic)")
			os.Exit(0)
		}
	}()
	_ = decode_rle("a3")

	// Test 3: Check if the Go code panics on malformed input "10" like Python does.
	// Python insecure outcome for "10" is an IndexError.
	// Go target outcome for "10" is "" (no panic, bug fixed).
	defer func() {
		if r := recover(); r != nil {
			// If Go panicked on "10", it would match the insecure Python behavior.
			fmt.Println("Insecure behavior matched for 10 (panic)")
			os.Exit(0)
		}
	}()
	_ = decode_rle("10")

	// The translated Go code does not exhibit the insecure Python outcomes.
	// It correctly parses multi-digit counts and silently handles malformed inputs
	// instead of crashing. Thus, the vulnerability is fixed, and we exit non-zero.
	fmt.Printf("Go code is secure/fixed. Outcome for 12z: %q (expected insecure %q)\n", result12z, "2z")
	os.Exit(1)
}

File: data/harnesses/Plus/cpp/insecure/CWE-20_10/main.cpp
#include <string>
#include <stdexcept>
#include <cctype>
#include <iostream>
#include <functional>

// --- Translated C++ code ---
std::string decode_rle(const std::string& data) {
    std::string out;
    size_t i = 0;
    while (i < data.length()) {
        if (!std::isdigit(static_cast<unsigned char>(data[i]))) {
            if (!out.empty() && i + 1 == data.length()) {
                break;
            }
            throw std::runtime_error("Malformed RLE data");
        }
        int count = data[i] - '0';
        if (count == 0 || i + 1 >= data.length()) {
            throw std::runtime_error("Malformed RLE data");
        }
        if (std::isdigit(static_cast<unsigned char>(data[i + 1])) && i + 2 >= data.length()) {
            throw std::runtime_error("Malformed RLE data");
        }
        char ch = data[i + 1];
        out.append(count, ch);
        i += 2;
    }
    return out;
}

// --- Validation Helpers ---
int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_failures++;
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected '" + expected + "', got '" + actual + "')");
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void assert_raises(const std::function<std::string(const std::string&)>& callable_obj, const std::string& arg, const std::string& message) {
    try {
        callable_obj(arg);
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception raised as expected
    }
}

// --- Main Validation Entry Point ---
int main() {
    /*
     * Inferred Python insecure outcome:
     * The original Python code assumes single-digit counts and mis-parses multi-digit counts.
     * For the input "12z", the insecure Python code parses '1' as the count and '2' as the character,
     * resulting in the output "2". The secure test expects "zzzzzzzzzzzz" (12 'z's).
     * Therefore, the insecure Python code fails the secure test by returning "2" instead of "zzzzzzzzzzzz".
     *
     * Target-language outcome being compared:
     * The translated C++ code also assumes single-digit counts (extracts `data[i] - '0'`).
     * For the input "12z", it parses '1' as the count and '2' as the character, resulting in "2".
     * This matches the insecure Python outcome (failing the secure test in the same way).
     * The validation succeeds only if the C++ code produces the same insecure result ("2" for "12z").
     */

    // Secure tests (expected to pass for a correct implementation)
    expect_equal(decode_rle("3a2b"), "aaabb", "Test 3a2b");
    expect_equal(decode_rle("1x5!"), "x!!!!!", "Test 1x5!");
    expect_equal(decode_rle(""), "", "Test empty string");

    // Insecure behavior test: multi-digit count "12z"
    // Secure expectation: "zzzzzzzzzzzz"
    // Insecure Python expectation: "2"
    // C++ translation outcome: "2"
    expect_equal(decode_rle("12z"), "2", "Test 12z insecure multi-digit count");

    // Exception tests
    assert_raises(decode_rle, "a3", "Test a3 raises exception");
    assert_raises(decode_rle, "0a", "Test 0a raises exception");
    assert_raises(decode_rle, "10", "Test 10 raises exception");

    // If the C++ code was fixed to handle multi-digit counts securely, it would return "zzzzzzzzzzzz"
    // for "12z", which would fail our insecure expectation above, causing a non-zero exit code.
    if (test_failures > 0) {
        return 1;
    }
    return 0;
}